server/src/main/java/org/elasticsearch/index/shard/ShardLongFieldRange.java

```diff
@@ -0,0 +1,137 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Class representing an (inclusive) range of {@code long} values in a field in a single shard.
+ */
+public class ShardLongFieldRange implements Writeable {
+
+    static final Version LONG_FIELD_RANGE_VERSION_INTRODUCED = Version.V_8_0_0;
+
+    /**
+     * Sentinel value indicating an empty range, for instance because the field is missing or has no values.
+     */
+    public static final ShardLongFieldRange EMPTY = new ShardLongFieldRange(Long.MAX_VALUE, Long.MIN_VALUE);
+
+    /**
+     * Sentinel value indicating the actual range may change in the future.
+     */
+    public static final ShardLongFieldRange MUTABLE = new ShardLongFieldRange(Long.MIN_VALUE, Long.MAX_VALUE);
+
+    /**
+     * Construct a new {@link ShardLongFieldRange} with the given (inclusive) minimum and maximum.
+     */
+    public static ShardLongFieldRange of(long min, long max) {
+        assert min <= max : min + " vs " + max;
+        return new ShardLongFieldRange(min, max);
+    }
+
+    private final long min, max;
+
+    private ShardLongFieldRange(long min, long max) {
+        this.min = min;
+        this.max = max;
+    }
+
+    /**
+     * @return the (inclusive) minimum of this range.
+     */
+    public long getMin() {
+        assert this != EMPTY && this != MUTABLE && min <= max: "must not use actual min of sentinel values";
+        return min;
+    }
+
+    /**
+     * @return the (inclusive) maximum of this range.
+     */
+    public long getMax() {
+        assert this != EMPTY && this != MUTABLE && min <= max : "must not use actual max of sentinel values";
+        return max;
+    }
+
+    @Override
+    public String toString() {
+        if (this == MUTABLE) {
+            return "MUTABLE";
+        } else if (this == EMPTY) {
+            return "EMPTY";
+        } else {
+            return "[" + min + "-" + max + "]";
+        }
+    }
+
+    public static ShardLongFieldRange readFrom(StreamInput in) throws IOException {
+        if (in.getVersion().before(LONG_FIELD_RANGE_VERSION_INTRODUCED)) {
+            // conservative treatment for BWC
+            return MUTABLE;
+        }
+
+        final byte type = in.readByte();
+        switch (type) {
+            case 1:
```

Comment:
nit: can we extract these constants to a static field?

Reply:
Yep, see a978aea54f52d357ab41bbc38d89fd26df3dd797.